src/main/java/frc/robot/subsystems/shooter/Shooter.java

```diff
@@ -0,0 +1,132 @@
+
+  ShooterIOInputsAutoLogged shooterInputs = new ShooterIOInputsAutoLogged();
+  double cmdTopSpeed;
+  double cmdBottomSpeed;
+  LoggedTunableNumber speedThreshold = new LoggedTunableNumber("Shooting speed threshold", 0.05);
+  LoggedTunableNumber firingTime = new LoggedTunableNumber("Firing Time", 0.5);
+
+  boolean isShooting = false;
+  public boolean isFiring = false;
+  public boolean intaking = false;
+  double feedCmd = 0.0;
+  private LoggedTunableNumber ltn = new LoggedTunableNumber("shoor fac", 4000);
+
+  public boolean isHolding() {
+    return holding;
+  }
+
+  public void setHolding(boolean holding) {
+    this.holding = holding;
+  }
+
+  boolean holding = false;
+
+  public void setFiringBlocked(boolean firingBlocked) {
+    this.firingBlocked = firingBlocked;
+  }
+
+  boolean firingBlocked = false;
+
+  double firingStartTime;
+
+  public Shooter(ShooterIO io, FeederIO feederIO) {
+    this.shooterIO = io;
+    this.feederIO = feederIO;
+    io.updateInputs(shooterInputs);
+    Logger.processInputs("Shooter/ShooterInputs", shooterInputs);
+    feederIO.updateInputs(feederInputs);
+    Logger.processInputs("Shooter/FeederInputs", feederInputs);
+  }
+
+  @Override
+  public void periodic() {
+    shooterIO.updateInputs(shooterInputs);
+    Logger.processInputs("Shooter/ShooterInputs", shooterInputs);
+    feederIO.updateInputs(feederInputs);
+    Logger.processInputs("Shooter/FeederInputs", feederInputs);
+    if (isShooting) {
+      shooterIO.setVelocity(cmdTopSpeed, cmdBottomSpeed);
+    } else {
+      shooterIO.setVelocity(0, 0);
+    }
+    if (!isFiring
+        && isShooting
+        && Math.abs((shooterInputs.velocity[0] - cmdTopSpeed) / cmdTopSpeed) < speedThreshold.get()
+        && Math.abs((shooterInputs.velocity[1] - cmdBottomSpeed) / cmdBottomSpeed)
+            < speedThreshold.get()
+        && !firingBlocked) {
+      isFiring = true;
+      firingStartTime = Timer.getFPGATimestamp();
+    }
+    // The open and closed should always be opposite, and anything else would be an electrical fault
+    if (intaking && (feederInputs.closedContact == feederInputs.openContact)) {
+      intaking = false;
+      // Shut down to prevent damage to ring
+      cmdFeeder(0);
+      cmdVel(0, 0);
+      holding = true;
+      System.out.println("Feeder limit switch disconnected!!");
+    }
+    if (intaking && feederInputs.closedContact) {
+      cmdFeeder(0);
+      cmdVel(0, 0);
+      holding = true;
+      intaking = false;
+    }
+    if (isFiring) {
+      cmdFeeder(-8000);
+      if (firingStartTime + firingTime.get() < Timer.getFPGATimestamp()) {
+        isFiring = false;
+        isShooting = false;
+        holding = false;
+        cmdFeeder(0);
+      }
+    }
+  }
+
+  public void cmdVel(double v1, double v2) {
+    shooterIO.setVelocity(v1, v2);
+  }
+
+  public void cmdVoltage(double voltage) {
+    shooterIO.setVoltage(voltage);
+  }
+
+  public void commandSpeed(double exitVel) {
```

Comment:
Why is this a different naming scheme from `cmdVel` and `cmdVoltage`? (This is `commandSpeed`, not `cmdSpeed`)